### PR TITLE
build go images with 1.21.7 for old release branches and drop go1.20 images build

### PR DESCRIPTION
**Pull request by cpanato**
#### What type of PR is this?

/kind cleanup

/kind feature

#### What this PR does / why we need it:

- build go images with 1.21.7 for old release branches and drop go1.20 images build

/assign @saschagrunert @xmudrii @puerco 
cc @liggitt @kubernetes/release-engineering 

#### Which issue(s) this PR fixes:

None

#### Does this PR introduce a user-facing change?

```release-note
build go images with 1.21.7 for old release branches and drop go1.20 images build
```

### CODE REVIEW

File: dependencies.yaml [hunk 1]
```diff
@@ -428,9 +406,6 @@ dependencies:
       match: OS_CODENAME\ \?=\ bullseye
     - path: images/build/cross/variants.yaml
       match: "OS_CODENAME: 'bullseye'"
-      # We keep them around for backwards compatibility and release branches.
-    - path: images/build/go-runner/variants.yaml
-      match: "OS_CODENAME: 'bullseye'"
```
Comment (liggitt) [inline dependencies.yaml:433]: does updating to go1.21 imply dropping bullseye? that is not obvious to me.

File: images/build/go-runner/variants.yaml [hunk 2]
```diff
@@ -15,11 +15,3 @@ variants:
     REVISION: '0'
     GO_VERSION: '1.21.7'
     DISTROLESS_IMAGE: 'static-debian12'
-  go1.20-bullseye:
```
Comment (liggitt) [inline images/build/go-runner/variants.yaml:18]: should this have changed to go1.21-bullseye?

File: dependencies.yaml (hunk 1 above)
Comment (cpanato) [reply dependencies.yaml:433]: that was for go1.20, in go1.21+ we are using bookworm

File: images/build/go-runner/variants.yaml (hunk 2 above)
Comment (cpanato) [reply images/build/go-runner/variants.yaml:18]: go runnuer on go1.21 is using bookworm, the release-1.29 is using go-runner `v2.3.1-go1.21.7-bookworm.0`

do you want to build go-runner for releases 1.28/27/26 with bullseye? or keep the same we are using for release-1.29?

File: dependencies.yaml (hunk 1 above)
Comment (liggitt) [reply dependencies.yaml:433]: is that *required* for go1.21, or just incidental? does that impact anything other than our build / test images? if it's just for CI / build, that's probably fine. if it's something we ship in the release, should we try to limit ourselves to one change at a time (changing go without changing os image as well)?

File: images/build/go-runner/variants.yaml (hunk 2 above)
Comment (liggitt) [reply images/build/go-runner/variants.yaml:18]: I... don't know enough to know for sure, but I like changing one variable at a time if possible, so I sort of expected to keep using bullseye for 1.26 / 1.27 / 1.28 but with go1.21

Comment (cpanato) [reply images/build/go-runner/variants.yaml:18]: ok, i will adjust that

**Pull request merged**